pkg/controller/common/terraform.go

```diff
@@ -61,3 +76,27 @@ func TerraformVariablesEnvironmentFromCredentials(credentials *alicloud.Credenti
 		TerraformVarAccessKeySecret: credentials.AccessKeySecret,
 	}
 }
+
+// IsResourceEmpty checks if the Terraformer state is not empty but contains no resources
+func IsResourceEmpty(tf terraformer.Terraformer) (bool, error) {
+	stateConfigMap, err := tf.GetState()
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+
+	var state tfState
+	if err := json.Unmarshal(stateConfigMap, &state); err != nil {
+		return false, err
+	}
+
+	for _, res := range state.Resources {
+		if res.Provider == TerraformProvider && len(res.Instances) > 0 {
```

Comment:
Is the provider check needed? Obviously, you can assume that the Alicloud extension only schedules Alicloud Terraform jobs, or?

Reply:
the provider check is necessary, because currently we have one `null provider` for workaround, see https://github.com/gardener/gardener-extension-provider-alicloud/blob/master/charts/internal/alicloud-infra/templates/main.tf#L102-L106.

it always will generate not empty instances in `tf.state` like:
```
{
    "mode": "managed",
    "type": "null_resource",
    "name": "outputs",
    "provider": "provider.null",
    "instances": [
        {
            "schema_version": 0,
            "attributes": {
                "id": "4543855011011788166",
                "triggers": {
                    "recompute": "outputs"
                }
            }
        }
    ]
}
```

Reply:
👍 Makes sense, thanks for the explanation!